File: examples/Lua/zmsg.lua
--
--  Ported zmsg object from zmsg.h
--
--  Author: Robert G. Jakabosky <bobby@sharedrealm.com>
--

require"zhelpers"

local zmq = require"zmq"

local setmetatable = setmetatable
local tinsert = table.insert
local tremove = table.remove
local tconcat = table.concat
local char = string.char
local tonumber = tonumber
local assert = assert
local stderr = io.stderr
local fprintf = fprintf

-- hex <-> binary conversion tables.
hex_to_byte_map = {}
byte_to_hex_map = {}
for i=0,255 do
    local byte = char(i)
    local hex = ('%02X'):format(i)
    -- byte to hex
    byte_to_hex_map[i] = hex
    byte_to_hex_map[byte] = hex
    -- hex to byte
    hex_to_byte_map[hex] = byte
end

--  --------------------------------------------------------------------------
--  Formats 17-byte UUID as 33-char string starting with '@'
--  Lets us print UUIDs as C strings and use them as addresses
--
local tmp_enc = {}
local function s_encode_uuid(data)
    assert(data:byte(1) == 0)
    tmp_enc[1] = '@'
    for i=2,17 do
        tmp_enc[i] = byte_to_hex_map[data:byte(i,i)]
    end
    return tconcat(tmp_enc, '', 1, 17)
end


--  --------------------------------------------------------------------------
--  Formats 17-byte UUID as 33-char string starting with '@'
--  Lets us print UUIDs as C strings and use them as addresses
--
local tmp_dec = {}
local function s_decode_uuid(uuidstr)
    assert(uuidstr:sub(1,1) == '@')
    tmp_dec[1] = '\0'
    local n=1
    for i=2,33,2 do
        n = n + 1
        tmp_dec[n] = hex_to_byte_map[uuidstr:sub(i,i+1)]
    end
    return tconcat(tmp_dec, '', 1, n)
end


local msg_mt = {}
msg_mt.__index = msg_mt

local function treverse(t)
    local len = #t
    local half = math.floor(len / 2)
    for i=1,half do
        t[len], t[i] = t[i], t[len]
        len = len - 1
    end
    return t
end

--  --------------------------------------------------------------------------
--  Constructor, sets initial body if provided
local function zmsg_new(body)
    return setmetatable({body}, msg_mt)
end


--  --------------------------------------------------------------------------
--  Compare two messages

function msg_mt:__eq(msg2)
    local len = #self
    -- different lengths?
    if len ~= #msg2 then return false end
    for i=1,len do
        -- different parts?
        if self[i] ~= msg2[i] then
            return false
        end
    end
    -- same message contents.
    return true
end


--  --------------------------------------------------------------------------
--  Duplicate message

function msg_mt:dup()
    local dup = zmsg_new()

    for i=1,#self do
        dup[i] = self[i]
    end
    return dup
end


--  --------------------------------------------------------------------------
--  Send message to socket
--  Destroys message after sending

function msg_mt:send(socket)
    assert(socket)
    local len = #self

    for i=len,1,-1 do
        local part = self[i]

        --  Unmangle 0MQ identities for writing to the socket
        if (#part == 33 and part:sub(1,1) == '@') then
            part = s_decode_uuid (part)
        end
        socket:send(part, (i > 1) and zmq.SNDMORE or 0)
        -- clear zmsg
        self[i] = nil
    end
end


--  --------------------------------------------------------------------------
--  Report size of message

function msg_mt:parts()
    return #self
end


--  --------------------------------------------------------------------------
--  Return pointer to message body, if any

function msg_mt:body()
    return self[1]
end


--  --------------------------------------------------------------------------
--  Set message body as copy of provided string
--  If message is empty, creates a new message body

function msg_mt:body_set(body)
    self[1] = body
end


--  --------------------------------------------------------------------------
--  Set message body using printf format
--  If message is empty, creates a new message body
--  Hard-coded to max. 255 characters for this simplified class

function msg_mt:body_fmt(fmt, ...)
    self:body_set(fmt:format(...))
end


--  --------------------------------------------------------------------------
--  Push message part to front of message parts

function msg_mt:push(part)
    self[#self + 1] = part
end


--  --------------------------------------------------------------------------
--  Pop message part off front of message parts

function msg_mt:pop()
    local len = #self
    local part = self[len]
    self[len] = nil
    return part
end


--  --------------------------------------------------------------------------
--  Append message part to end of message parts

function msg_mt:append(part)
    tinsert(self, 1, part)
end


--  --------------------------------------------------------------------------
--  Return pointer to outer message address, if any

function msg_mt:address()
    return self[#self]
end


--  --------------------------------------------------------------------------
--  Wraps message in new address envelope
--  If delim is not null, creates two-part envelope
--  Call this _after_ zmq_body_set(), not before

function msg_mt:wrap(address, delim)
    assert(address)

    --  Push optional delimiter and then address
    if delim then
        self:push(delim)
    end
    self:push(address)
end


--  --------------------------------------------------------------------------
--  Unwraps outer message envelope and returns address
--  Discards empty message part after address, if any

function msg_mt:unwrap()
    assert(self)

    local address = self:pop()
    -- check for deliminator
    local peek = self:address()
    if (peek and #peek == 0) then
        -- pop deliminator
        self:pop()
    end
    return address
end


--  --------------------------------------------------------------------------
--  Save message to a file

function msg_mt:save(file)
    assert(file)

    for i=#self,1,-1 do
        local part = self[i]
        -- prefix each message part with: "<8 digit hex number>:"
        file:write(("%08X:"):format(#part))
        file:write(part)
    end
end


--  --------------------------------------------------------------------------
--  Dump message to stderr, for debugging and tracing

function msg_mt:dump()
    fprintf(stderr, "--------------------------------------\n")
    for i=#self,1,-1 do
        local data = self[i]

        --  Dump the message as text or binary
        local is_text = true
        for i=1,#data do
            local c = data:byte(i)
            if (c < 32 or c > 127) then
                is_text = false
                break
            end
        end

        fprintf(stderr, "[%03d] ", #data)
        if (is_text) then
            stderr:write(data)
        else
            for i=1,#data do
                fprintf(stderr, "%02X", data:byte(i))
            end
        end
        fprintf(stderr, "\n")
    end
    stderr:flush()
end

module(...)

-- export zmsg_new()
_M.new = zmsg_new


--  --------------------------------------------------------------------------
--  Receive message from socket
--  Creates a new message and returns it
--  Blocks on recv if socket is not ready for input

function recv(socket, flags)
    assert(socket)
    local self = zmsg_new()

    while true do
        local data, err = socket:recv(flags)
        if not data then
            return treverse(self), err
        end
        --  We handle 0MQ UUIDs as printable strings
        if (#data == 17 and data:byte(1) == 0) then
            --  Store message part as string uuid
            data = s_encode_uuid (data)
        end
        self[#self + 1] = data

        if (socket:getopt(zmq.RCVMORE) == 0) then
            break      --  Last message part
        end
    end
    return treverse(self)
end


--  --------------------------------------------------------------------------
--  Load message from file
--  Creates a new message and returns as many parts as can be read.

function load(file)
    assert(file)
    local self = zmsg_new ()

    while true do
        -- Read part prefix
        local size = file:read(9)
        if not size then break end
        -- decode part size to number
        size = tonumber(size:sub(1,8), 16)
        -- Read part from file
        local part = file:read(size)
        if not part then break end
        self[#self + 1] = part
    end
    return treverse(self)
end

File: examples/PHP/zmsg.php
<?php
/*  =========================================================================
    zmsg.php

    Multipart message class for example applications.

    Follows the ZFL class conventions and is further developed as the ZFL
    zfl_msg class.  See http://zfl.zeromq.org for more details.

    -------------------------------------------------------------------------
    Copyright (c) 1991-2010 iMatix Corporation <www.imatix.com>
    Copyright other contributors as noted in the AUTHORS file.

    This file is part of the ZeroMQ Guide: http://zguide.zeromq.org

    This is free software; you can redistribute it and/or modify it under the
    terms of the GNU Lesser General Public License as published by the Free
    Software Foundation; either version 3 of the License, or (at your option)
    any later version.

    This software is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABIL-
    ITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General
    Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program. If not, see <http://www.gnu.org/licenses/>.
    =========================================================================

   @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
*/

class zmsg
{
    /**
     * Store the parts of the message
     *
     * @var array
     */
    private $_parts = array();

    /**
     * Socket to send and receive via
     *
     * @var ZMQSocket
     */
    private $_socket;

    /**
     * Constructor, accepts optional socket for sending/receiving.
     *
     * @param ZMQSocket $socket
     */
    public function __construct($socket = null)
    {
        $this->_socket = $socket;
    }

    /**
     * Formats 17-byte UUID as 33-char string starting with '@'
     * Lets us print UUIDs as C strings and use them as addresses
     *
     * @param  string $data
     * @return string
     */
    public function s_encode_uuid($data)
    {
        return "@" . bin2hex($data);
    }

    /**
     * Format the hex string back into a packed int.
     *
     * @param  string $data
     * @return string
     */
    public function s_decode_uuid($data)
    {
        return pack("H*", substr($data, 1));
    }

    /**
     * Set the internal socket to use for sending or receiving.
     *
     * @param  ZMQSocket $socket
     * @return Zmsg
     */
    public function set_socket(ZMQSocket $socket)
    {
        $this->_socket = $socket;

        return $this;
    }

    /**
     *  Receive message from socket
     *  Creates a new message and returns it
     *  Blocks on recv if socket is not ready for input
     *
     * @throws Exception if no socket present
     * @return Zmsg
     */
    public function recv()
    {
        if (!isset($this->_socket)) {
            throw new Exception("No socket supplied");
        }
        $this->_parts = array();
        while (true) {
            $this->_parts[] = $this->_socket->recv();
            if (!$this->_socket->getSockOpt(ZMQ::SOCKOPT_RCVMORE)) {
                break;
            }
        }

        return $this;
    }

    /**
     * Send message to socket. Destroys message after sending.
     *
     * @throws Exception if no socket present
     * @param  boolean   $clear
     * @return Zmsg
     */
    public function send($clear = true)
    {
        if (!isset($this->_socket)) {
            throw new Exception("No socket supplied");
        }
        $count = count($this->_parts);
        $i = 1;
        foreach ($this->_parts as $part) {
            $mode = $i++ == $count ? null : ZMQ::MODE_SNDMORE;
            $this->_socket->send($part, $mode);
        }
        if ($clear) {
            unset($this->_parts);
            $this->_parts = array();
        }

        return $this;
    }

    /**
     * Report size of message
     *
     * @return int
     */
    public function parts()
    {
        return count($this->_parts);
    }

    /**
     * Return the last part of the message
     *
     * @return string
     */
    public function last()
    {
        return $this->_parts[count($this->_parts)-1];
    }

    /**
     * Set the last part of the message
     *
     * @param string $set
     */
    public function set_last($set)
    {
        $this->_parts[count($this->_parts)-1] = $set;
    }

    /**
     * Return the body
     *
     * @return string
     */
    public function body()
    {
        return $this->_parts[count($this->_parts) -1];
    }

    /**
     * Set message body to provided string.
     *
     * @param  string $body
     * @return Zmsg
     */
    public function body_set($body)
    {
        $pos = count($this->_parts);
        if ($pos > 0) {
            $pos = $pos - 1;
        }
        $this->_parts[$pos] = $body;

        return $this;
    }

    /**
     * Set message body using printf format
     *
     * @return void
     */
    public function body_fmt()
    {
        $args = func_get_args();
        $this->body_set(vsprintf(array_shift($args), $args));

        return $this;
    }

    /**
     * Push message part to front
     *
     * @param  string $part
     * @return void
     */
    public function push($part)
    {
        array_unshift($this->_parts, $part);
    }

    /**
     * Pop message part off front of message parts
     *
     * @return string
     * @author Ian Barber
     */
    public function pop()
    {
        return array_shift($this->_parts);
    }

    /**
     * Return the address of the message
     *
     * @return void
     * @author Ian Barber
     */
    public function address()
    {
        $address = count($this->_parts) ? $this->_parts[0] : null;

        return (strlen($address) == 17 && $address[0] == 0) ? $this->s_encode_uuid($address) : $address;
    }

    /**
     * Wraps message in new address envelope.
     * If delim is not null, creates two part envelope.
     *
     * @param  string $address
     * @param  string $delim
     * @return void
     */
    public function wrap($address, $delim = null)
    {
        if ($delim !== null) {
            $this->push($delim);
        }
        if ($address[0] == '@' && strlen($address) == 33) {
            $address = $this->s_decode_uuid($address);
        }
        $this->push($address);

        return $this;
    }

    /**
     * Unwraps outer message envelope and returns address
     * Discards empty message part after address, if any
     *
     * @return void
     * @author Ian Barber
     */
    public function unwrap()
    {
        $address = $this->pop();
        if (!$this->address()) {
            $this->pop();
        }

        return $address;
    }

    /**
     * Dump the contents to a string, for debugging and tracing.
     *
     * @return string
     */
    public function __toString()
    {
        $string = "--------------------------------------" . PHP_EOL;
        foreach ($this->_parts as $index => $part) {
            $len = strlen($part);
            if ($len == 17 && $part[0] == 0) {
                $part = $this->s_encode_uuid($part);
                $len = strlen($part);
            }
            $string .= sprintf ("[%03d] %s %s",  $len, $part, PHP_EOL);
        }

        return $string;
    }

    /**
     * Run a self test of the Zmsg class.
     *
     * @return boolean
     * @todo See if assert returns
     */
    public static function test()
    {
        $result = true;
        $context = new ZMQContext();
        $output = new ZMQSocket($context, ZMQ::SOCKET_DEALER);
        $output->bind("inproc://zmsg_selftest");

        $input = new ZMQSocket($context, ZMQ::SOCKET_ROUTER);
        $input->connect("inproc://zmsg_selftest");

        //  Test send and receive of single-part message
        $zmsgo = new Zmsg($output);
        $zmsgo->body_set("Hello");
        $result &= assert($zmsgo->body() == "Hello");
        $zmsgo->send();

        $zmsgi = new Zmsg($input);
        $zmsgi->recv();
        $result &= assert($zmsgi->parts() == 2);
        $result &= assert($zmsgi->body() == "Hello");
        echo $zmsgi;

        //  Test send and receive of multi-part message
        $zmsgo = new Zmsg($output);
        $zmsgo->body_set("Hello");
        $zmsgo->wrap("address1", "");
        $zmsgo->wrap("address2");
        $result &= assert($zmsgo->parts() == 4);
        echo $zmsgo;
        $zmsgo->send();

        $zmsgi = new Zmsg($input);
        $zmsgi->recv();
        $result &= assert($zmsgi->parts() == 5);
        $zmsgi->unwrap();
        $result &= assert($zmsgi->unwrap() == "address2");

        $zmsgi->body_fmt("%s%s", 'W', "orld");
        $result &= assert($zmsgi->body() == "World");

        //  Pull off address 1, check that empty part was dropped
        $zmsgi->unwrap();
        $result &= assert($zmsgi->parts() == 1);

        //  Check that message body was correctly modified
        $part = $zmsgi->pop();
        $result &= assert ($part == "World");
        $result &= assert($zmsgi->parts() == 0);

        // Test load and save
        $zmsg = new Zmsg();
        $zmsg->body_set("Hello");
        $zmsg->wrap("address1", "");
        $zmsg->wrap("address2");
        $result &= assert($zmsg->parts() == 4);
        $fh = fopen(sys_get_temp_dir() . "/zmsgtest.zmsg", 'w');
        $zmsg->save($fh);
        fclose($fh);
        $fh = fopen(sys_get_temp_dir() . "/zmsgtest.zmsg", 'r');
        $zmsg2 = new Zmsg();
        $zmsg2->load($fh);
        assert($zmsg2->last() == $zmsg->last());
        fclose($fh);
        $result &= assert($zmsg2->parts() == 4);
        echo ($result ? "OK" : "FAIL"), PHP_EOL;

        return $result;
    }

    /**
     * Save a msg to a file
     *
     * @param  filehandle $fh
     * @return this
     */
    public function save($fh)
    {
        foreach ($this->_parts as $part) {
            fwrite($fh, chr(strlen($part)));
            if (strlen($part) > 0) {
                fwrite($fh, $part);
            }
        }

        return $this;
    }

    /**
     * Load a message saved with the save function
     *
     * @param  filehandle $fh
     * @return this
     */
    public function load($fh)
    {
        while (!feof($fh) && $size = fread($fh, 1)) {
            $this->_parts[] = ord($size) > 0 ? fread($fh, ord($size)) : '';
        }

        return $this;
    }
}
